src/mem/mem_scheduler.cc

```diff
@@ -0,0 +1,634 @@
+    avgWQDelay.precision(2);
+    avgRespDelay.precision(2);
+    avgRQDelay = totalRQDelay / readReqs;
+    avgWQDelay = totalWQDelay / writeReqs;
+    avgRespDelay = totalRespDelay / readReqs;
+}
+
+bool
+MemScheduler::handleRequest(PortID portId, PacketPtr pkt)
+{
+      ///////////////////////////////////////////////////////////////////////
+     /////////////////// TODO: adjust the schedule freq ////////////////////
+    ///////////////////////////////////////////////////////////////////////
+    DPRINTF(MemScheduler, "handleRequest: Got request from cpuPort[%d]\n", portId);
+    auto queue = find_if(requestQueues.begin(), requestQueues.end(), [portId](RequestQueue &obj){return obj.cpuPortId == portId;});
+    if (queue->blocked(pkt->isRead() && !pkt->isWrite())){
+        if (pkt->isWrite() && !unifiedQueue){
+            DPRINTF(MemScheduler, "handleRequest: Blocking write request\n");
+            queue->sendWriteRetry = true;
+        }
+        if (pkt->isRead() || (unifiedQueue && pkt->isWrite())){
+            DPRINTF(MemScheduler, "handleRequest: Blocking read request\n");
+            queue->sendReadRetry = true;
+        }
+        return false;
+    }
+    if (pkt->isRead()){
+        stats.readReqs++;
+        entryTimes[pkt] = curTick();
+        DPRINTF(MemScheduler, "handleRequest: Pushing read request to readQueues[%d], size = %d\n", portId, queue->readQueue.size());
+        queue->push(pkt);
+        DPRINTF(MemScheduler, "handleRequest: Pushed read request to readQueues[%d], size = %d\n", portId, queue->readQueue.size());
+    } else{
+        stats.writeReqs++;
+        entryTimes[pkt] = curTick();
+        DPRINTF(MemScheduler, "handleRequest: Pushing write request to writeQueues[%d], size = %d\n", portId, queue->writeQueue.size());
+        queue->push(pkt);
+        DPRINTF(MemScheduler, "handleRequest: Pushed write request to writeQueues[%d], size = %d\n", portId, queue->writeQueue.size());
+    }
+    if (!nextReqEvent.scheduled()){
+        DPRINTF(MemScheduler, "handleRequest: Scheduling nextReqEvent in handleRequest\n");
+        schedule(nextReqEvent, curTick() + 100);
+    }
+    return true;
+}
+
+void
+MemScheduler::processNextReqEvent(){
+
+    uint64_t minCheck = -1;
+    RequestQueue *queue;
+    DPRINTF(MemScheduler, "processNextReqEvent: Finding the least recently visited non-empty queue\n");
+    for (auto &it : requestQueues){
+        if ((it.timesChecked < minCheck) && (!it.emptyRead() || it.serviceWrite())){
+            minCheck = it.timesChecked;
+            queue = &it;
+        }
+    }
+
+    stats.totalArbitrations++;
+    DPRINTF(MemScheduler, "processNextReqEvent: Least recently visited queue found, cpuPortId = %d, (timesChecked = %d) == (minChecked = %d), readQueue.size = %d, writeQueue.size = %d, emptyRead = %d, serviceWrite = %d\n", queue->cpuPortId, queue->timesChecked, minCheck, queue->readQueue.size(), queue->writeQueue.size(), queue->emptyRead(), queue->serviceWrite());
```

Comment:
This line is ridiculous. 80 chars might be small, but there has to be some limit. You really should be following gem5's style guideline to a "T" even if you're working on research.

How can anyone read/understand this line?